### PR TITLE
M4: fix: store JSON null for empty response bodies in idempotencyfilter

### PR DESCRIPTION
**Pull request by BanerjeeAgniva**
# Pull Request

## Summary
Fix `IdempotencyFilter` so empty response bodies (e.g. 202/204 from M8's scan endpoints) persist correctly to the `NOT NULL jsonb` `idempotency_keys.response_body` column, restoring HTTP-level replay protection on those routes.

---

## What Changed
- On write: store the JSON `null` literal instead of an empty string when a handler returns no content (empty string isn't valid `jsonb`).
- On replay: skip writing a body (and the JSON content-type) when the cached body is the `null` sentinel, faithfully reproducing an empty response.

---

## Why This Change?
`response_body` is `JSONB NOT NULL`. The filter stored the raw body verbatim, so an empty-body POST tried to insert `""` — which Postgres rejects as `invalid input syntax for type json`. The insert failed silently on every empty-body POST: the request still succeeded, but no idempotency record was persisted, so replay protection didn't apply to those routes. JSON-body endpoints (e.g. booking) were never affected.

---

## Testing
- [x] Tested locally
- [x] Edge cases considered
- [x] No breaking changes introduced

### Test Evidence
Replaying an M8 202 scan POST with the same `Idempotency-Key` now returns the cached 202 with an empty body and `X-Idempotency-Replayed: true`; the record is present in `idempotency_keys` with `response_body = null`.

---

### Areas Affected
- [x] Backend
- [x] Database

### Modules Affected
- [x] M4 — Orders (surfaced by M8 — Barcode)

---

## Risks / Concerns
- Low. Single-file change; JSON-body endpoints unchanged. The `null` literal is used as an in-band sentinel — a handler that deliberately returns a JSON `null` payload would replay as an empty body, which no current endpoint does.

---

## Checklist
- [x] Code follows project conventions (package layout, no cross-module internal imports)
- [x] Self-review completed
- [x] Append-only audit trail preserved (no mutations to scan/manifest/role-change records)
- [x] No sensitive data or credentials exposed
- [x] Documentation updated if behaviour changed
- [x] Ready for review

### CODE REVIEW

File: orders/src/main/java/com/oneday/orders/api/filter/IdempotencyFilter.java [hunk 1]
```diff
@@ -97,6 +97,14 @@ public class IdempotencyFilter extends OncePerRequestFilter {
...
+    /**
+     * Stand-in stored for an empty response body. {@code response_body} is a {@code NOT NULL jsonb}
+     * column, and an empty string is not valid JSON — a handler returning no content (e.g. a 202/204)
+     * would otherwise fail the insert with {@code invalid input syntax for type json}. We store the
+     * JSON {@code null} literal and reproduce an empty body on replay.
+     */
+    private static final String JSON_NULL_LITERAL = "null";
+
     // Deliberately a bare ObjectMapper (no custom serializers) so the fingerprint is
     // unaffected by application-level Jackson configuration on the injected objectMapper.
```
Comment (BanerjeeAgniva) [inline orders/src/main/java/com/oneday/orders/api/filter/IdempotencyFilter.java:109]: give me an example on what bug this solves? real business scenario 

what was before vs what is now 

explain end to end , end to end as in write first mile / last mile , write what has happenned till now , what problem you are solving , what goes on after this

Comment (BanerjeeAgniva) [reply orders/src/main/java/com/oneday/orders/api/filter/IdempotencyFilter.java:109]: ### What bug this solves

**One line:** on the scan endpoints that reply with no body (`202`/`204`), the idempotency record was silently *never getting stored*, so those endpoints had **zero replay protection** — a retried scan ran twice.

---

**Before vs now**

- **Before:** on a miss we ran the handler, then did `record.setResponseBody(responseBody)` with `responseBody = ""`. `idempotency_keys.response_body` is a `NOT NULL jsonb` column, and `""` is not valid JSON. The `INSERT` blew up with `invalid input syntax for type json`. That throw was swallowed by the `catch (Exception e)` at the bottom of `doFilterInternal` (the block meant for "a concurrent request already inserted this key") and logged as a harmless-looking `warn`. Net effect: the key row was **never written**. Every retry with the same `Idempotency-Key` was treated as a fresh miss and re-executed the handler.
- **Now:** an empty body is stored as the JSON `null` literal (`"null"`), which is valid `jsonb`, so the row persists. On replay we recognise that sentinel and reproduce an **empty** body (no `Content-Type`, no write) instead of echoing the literal string `null` back to the client.

---

**End to end — where this bites in the real chain**

Take the first-mile / hub leg. A delivery associate's phone or a hub scanner posts a scan — `VAN_LOAD`, pickup scan, `DA_TO_VAN`, etc. — to M8's scan-ingest endpoint. Those endpoints ack with `202/204` and **no body** (there's nothing to return; the scan is just appended to the ledger).

1. **What's happened until now:** parcel is booked (M4), routed onto a van loop (M6). The DA scans it onto the van. Phones on the road have flaky signal, so the client **retries** the POST — that's exactly why we require an `Idempotency-Key` header: the retry carries the *same* key so the second call should be a no-op replay.
2. **The problem being solved:** because the record was never stored (empty-body insert failed), the retry didn't replay — it **re-ran the handler**. The scan ledger is **append-only** (a core invariant — see CLAUDE.md), so that second run writes a *second* `VAN_LOAD` scan for the same parcel. You now have a duplicate custody event you can't delete: phantom parcel counts on the van manifest, a corrupted custody state in M6, and downstream SLA/reconciliation (M10/M11) reading a ledger that says one parcel was loaded twice. With this fix the retry hits a stored row and replays the `204` — the handler runs exactly once.
3. **What goes on after:** the van drives the loop, the parcel gets scanned at each custody handoff and finally `VAN_UNLOAD` / last-mile `DROP_COLLECTED` → `DROPPED`. Every one of those hops is the same empty-body scan POST over the same flaky mobile network, so the same retry/dedup guarantee has to hold end to end. This change is what restores that guarantee on the no-content routes — the JSON-body endpoints (B2C/B2B booking, which return a full response) were never affected.

Scope note: purely the empty-body case. Endpoints that return a real JSON body were already storing and replaying correctly; this doesn't touch them.
